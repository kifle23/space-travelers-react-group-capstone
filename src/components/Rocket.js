const Rocket = (rocket) => {
  const { name, img, desc } = rocket;
  return (
    <div className=" d-flex-space g-10 my-20">
      <img src={img} className="logo" alt="rocket" />
      <div className="rocket-details">
        <h2 className="rocket-name">{name}</h2>
        <p className="rocket-desc">{desc}</p>
        <button type="button" className="rocket-reserve btn">Reserve Rocket</button>
      </div>
    </div>
  );
};

export default Rocket;
