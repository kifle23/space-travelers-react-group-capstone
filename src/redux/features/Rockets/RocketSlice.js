import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  loading: false,
  rocket: [],
  error: '',
};
const url = 'https://api.spacexdata.com/v4/rockets';
export const fetchRockets = createAsyncThunk('rocket/fetchrockets', () => axios.get(url).then((res) => res.data));

const RocketSlice = createSlice({
  name: 'rocket',
  initialState,
  extraReducers: (builder) => {
    builder.addCase(fetchRockets.pending, (state) => ({
      ...state, loading: true,
    }));
    builder.addCase(fetchRockets.fulfilled, (state, action) => ({
      ...state,
      loading: false,
      rocket: action.payload,
      error: '',
    }));

    builder.addCase(fetchRockets.rejected, (state, action) => ({
      ...state,
      loading: false,
      rocket: [],
      error: action.error.message,
    }));
  },
});

export default RocketSlice.reducer;
