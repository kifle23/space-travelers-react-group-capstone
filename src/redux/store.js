import { configureStore } from '@reduxjs/toolkit';
import rocketsReducer from './features/Rockets/RocketSlice';

const store = configureStore({
  reducer: {
    rockets: rocketsReducer,
  },
});

const selectRockets = (store) => store.rockets;
export { selectRockets };
export default store;
