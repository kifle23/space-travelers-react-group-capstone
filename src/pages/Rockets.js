import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Rocket from '../components/Rocket';
import { fetchRockets } from '../redux/features/Rockets/RocketSlice';

const Rockets = () => {
  const rockets = useSelector((store) => store.rockets);
  const dispatch = useDispatch();
  const lengthRock = rockets.rocket.length;
  useEffect(() => {
    if (lengthRock > 0) return;
    dispatch(fetchRockets());
  }, [dispatch, lengthRock]);
  return (
    <div className="container">
      {rockets.loading && <div>loading</div>}
      {!rockets.loading && rockets.error ? (
        <div>
          Error
          {rockets.error}
        </div>
      ) : null}
      {!rockets.loading && rockets.rocket.length ? (
        rockets.rocket.map((rocket) => (
          <Rocket
            key={rocket.id}
            name={rocket.name}
            img={rocket.flickr_images}
            id={rocket.id}
            desc={rocket.description}
            reserved={rocket.reserved}
          />
        ))
      ) : null}
    </div>
  );
};

export default Rockets;
