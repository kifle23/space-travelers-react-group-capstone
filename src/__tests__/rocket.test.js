import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../redux/store';
import Rockets from '../pages/Rockets';
import Rocket from '../components/Rocket';

describe('A test for Rockets component', () => {
  it('Check whether rocket component renders', () => {
    const rocket = {
      id: 'heloo',
      name: 'Commercial Resupply Services',
      desc:
        'Commercial Resupply Services (CRS) are a series of contracts awarded by NASA from 2008â€“2016 for delivery of cargo and supplies to the International Space Station (ISS) on commercially operated spacecraft. The first CRS contracts were signed in 2008 and awarded $1.6 billion to SpaceX for 12 cargo transport missions and $1.9 billion to Orbital Sciences for 8 missions, covering deliveries to 2016. In 2015, NASA extended the Phase 1 contracts by ordering an additional three resupply flights from SpaceX and one from Orbital Sciences. After additional extensions late in 2015, SpaceX is currently scheduled to have a total of 20 missions and Orbital 10. SpaceX began flying resupply missions in 2012, using Dragon cargo spacecraft launched on Falcon 9 rockets from Space Launch Complex 40 at Cape Canaveral Air Force Station, Cape Canaveral, Florida. Orbital Sciences began deliveries in 2013 using Cygnus spacecraft launched on the Antares rocket from Launch Pad 0A at the Mid-Atlantic Regional Spaceport (MARS), Wallops Island, Virginia. A second phase of contracts (known as CRS2) were solicited and proposed in 2014. They were awarded in January 2016 to Orbital ATK, Sierra Nevada Corporation, and SpaceX, for cargo transport flights beginning in 2019 and expected to last through 2024.',
      reserved: false,
    };
    const renderedRockets = renderer.create(
      <Provider store={store}>
        <Rocket rocket={rocket} />
      </Provider>,
    );

    expect(renderedRockets).toMatchSnapshot();
  });

  it('Check if it renders correctly', () => {
    const renderedRockets = renderer.create(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );

    expect(renderedRockets).toMatchSnapshot();
  });
});
