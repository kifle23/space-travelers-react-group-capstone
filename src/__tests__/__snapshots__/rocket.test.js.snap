// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A test for Rockets component Check if it renders correctly 1`] = `
<div
  className="container"
/>
`;

exports[`A test for Rockets component Check whether rocket component renders 1`] = `
<div
  className=" d-flex-space g-10 my-20"
>
  <img
    alt="rocket"
    className="logo"
  />
  <div
    className="rocket-details"
  >
    <h2
      className="rocket-name"
    />
    <p
      className="rocket-desc"
    >
       
    </p>
    <button
      className="rocket-reserve btn"
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;
